### ISSUE TITLE
Refactor tlstest.py

### ISSUE THREAD

**Issue by FiachAntaw** [tomato42/tlslite-ng#80]
The current integration testing code could benefit from some refactoring. The auto-numbering is a massive improvement, but there's still some minor annoyances:

* The output isn't intuitive, particularly for the fault-injection tests. Printing 'pass' or 'ok' after each successful test would be an improvement, but isn't really feasible without changing how output is handled (at the moment both the client and server share stdout).
  * Fault injection tests indicate success with strings like 'Good Fault bad padding' which is obscure and could be interpreted as meaning the test failed, particularly since most tests don't produce any output at all.
* Several tests have different names on the client and server (e.g. test 0 is called 'anonymous handshake' on the client and 'Anonymous server handshake' on the server).
* Typically only one end of the connection crashes, and the other then falls over with an unrelated error ('Connection Refused' or 'TLSAbruptCloseError'). It's generally obvious which end failed, but it's still less than ideal.
* If testing is interrupted (e.g. with SIGINT/ctrl-C) or something causes the tlstest client to crash, the server can be left running which causes subsequent `make test` runs to fail ('Address already in use'). The exact circumstances for triggering this are unclear, beyond getting (un)lucky when you abort the test.

One option that comes to mind is to rewrite the tests to use the async API, which would remove the need for separate client and server processes. Testing could then be managed with the unittest module, to be consistent with the unit tests. Seem reasonable?

Also, a couple of minor notes related to testing in general:
* Running unit tests on Python 2 requires the 'mock' module to be installed, this doesn't appear to be in the list of requirements in CONTRIBUTING.md. Likewise, epydoc (and the associated modifications) are also missing from the list.
* The current `make test` target doesn't seem particularly useful given that it's undocumented and tests the installed copy of tlslite instead of the local copy. It also doesn't run unit tests and only runs the test on the default python. Perhaps the 'test' and 'test-local' targets should be removed, and 'test-dev' split up so that 'test' can invoke the unit/integration tests without running the linter/coverage checker.

**Comment by tomato42** [comment]
First of all, I don't like the whole `tlstest.py` - I consider it obsolete and on life support only. I would like to see it removed completely. As you've mentioned, it has a lot of problems.

Unfortunately, this is the only test coverage we have for [`tlslite/tlsconnection.py`](https://github.com/tomato42/tlslite-ng/blob/master/tlslite/tlsconnection.py) and [`tlslite/tlsrecordlayer.py`](https://github.com/tomato42/tlslite-ng/blob/master/tlslite/tlsrecordlayer.py) module. As long as that's the case, it needs to stay.

That being said, connection tests as unittest modules is definitely something I'd like to see in the future.

> Running unit tests on Python 2 requires the 'mock' module to be installed,
> this doesn't appear to be in the list of requirements in CONTRIBUTING.md.

aah, yes, I'm missing 'unittest' and `unittest2` dependency in CONTRIBUTING.md, will do a PR for that

> The current make test target doesn't seem particularly useful given that it's
> undocumented and tests the installed copy of tlslite instead of the local copy.

It's documented in the "Installation/Manual" section of the README:

```
From the distribution's directory, run: make test 
```

> It also doesn't run unit tests and only runs the test on the default python.

That was its purpose - to sanity test the installed version of library. I have to agree that it is rather uncommon and thus confusing usage. It's a leftover from times where there were no other tests.

The default python version is intentional - the `install` target will install it on default python too

> Perhaps the 'test' and 'test-local' targets should be removed, and 'test-dev'
> split up so that 'test' can invoke the unit/integration tests without running the
> linter/coverage checker.

unittests probably should be moved to `setup.py` (I have a hunch that it will be necessary for Windows support anyway)

`test` and `test-local` probably should have names switched, but I don't think it should have its names switched - those are the most complex tests and ones that are most irritating to run manually when developing.

**Comment by FiachAntaw** [comment]
>That being said, connection tests as unittest modules is definitely something I'd like to see in the future.

Agreed. There's definite value in maintaining some form of integration testing, if only to provide a safety net in case the unit tests overlook something critical.

>aah, yes, I'm missing 'unittest' and unittest2 dependency in CONTRIBUTING.md, will do a PR for that

Thanks.

>It's documented in the "Installation/Manual" section of the README

Ah, I must have overlooked that somehow. In context it does make sense to do it that way.

Should I close this issue or leave it open until tlstest.py is dead and gone?

**Comment by tomato42** [comment]
Let's leave it open - it does document the future intent.
